### PR TITLE
Resurrect: rustc_llvm: Add a -Z `print-codegen-stats` option to expose LLVM statistics.

### PR DESCRIPTION
**Pull request by khei4**
This resurrects PR https://github.com/rust-lang/rust/pull/104000, which has sat idle for a while. And I want to see the effect of stack-move optimizations on LLVM (like https://reviews.llvm.org/D153453) :).

I have applied the changes requested by @oli-obk and @nagisa  https://github.com/rust-lang/rust/pull/104000#discussion_r1014625377 and https://github.com/rust-lang/rust/pull/104000#discussion_r1014642482 in the latest commits.

r? @oli-obk 

-----

LLVM has a neat [statistics](https://llvm.org/docs/ProgrammersManual.html#the-statistic-class-stats-option) feature that tracks how often optimizations kick in. It's very handy for optimization work. Since we expose the LLVM pass timings, I thought it made sense to expose the LLVM statistics too.

-----
(Edit: fix broken link
(Edit2: fix segmentation fault and use malloc 

If `rustc` is built with 
```toml
[llvm]
assertions = true
```
Then you can see like 
```
rustc +stage1 -Z print-codegen-stats -C opt-level=3  tmp.rs
===-------------------------------------------------------------------------===
                          ... Statistics Collected ...
===-------------------------------------------------------------------------===
         3 aa                           - Number of MayAlias results
       193 aa                           - Number of MustAlias results
       531 aa                           - Number of NoAlias results
...
```

And the current default build emits only 
```
$ rustc +stage1 -Z print-codegen-stats -C opt-level=3  tmp.rs
===-------------------------------------------------------------------------===
                          ... Statistics Collected ...
===-------------------------------------------------------------------------===
$ 
```
This might be better to emit the message to tell assertion flag necessity, but now I can't find how to do that...

### CODE REVIEW

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 1]
```diff
@@ -111,9 +112,24 @@ extern "C" void LLVMRustSetNormalizedTarget(LLVMModuleRef M,
...
-  TimerGroup::printAll(OS);
+extern "C" const char *LLVMRustPrintPassTimings(void) {
+  std::string buf;
+  raw_string_ostream SS(buf);
+  TimerGroup::printAll(SS);
+  SS.flush();
+  char* CStr = (char*) malloc((buf.length() + 1) * sizeof(char));
+  strcpy(CStr, buf.c_str());
+  return CStr;
+}
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:123]: I would do something like this:
```suggestion
extern "C" const char *LLVMRustPrintPassTimings(size_t *Len) {
  std::string buf;
  raw_string_ostream SS(buf);
  TimerGroup::printAll(SS);
  SS.flush();
  *Len = bug.length();
  char *CStr = malloc(*Len);
  memcpy(CStr, buf.c_str(), *Len);
  return CStr;
}
```
And then use from_raw_parts on the Rust side to create he `[u8]` slice.

File: compiler/rustc_codegen_llvm/src/lib.rs [hunk 2]
```diff
@@ -177,9 +177,32 @@ impl WriteBackendMethods for LlvmCodegenBackend {
...
+        println!("{}", msg);
+    }
+    fn print_statistics(&self) {
+        let msg = unsafe {
+            let cstr = llvm::LLVMRustPrintStatistics();
+            if cstr.is_null() {
+                "failed to get stats".into()
+            } else {
+                let stats = CStr::from_ptr(cstr).to_bytes();
+                let stats = String::from_utf8_lossy(stats).to_string();
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/lib.rs:200]: We expect stats to be valid UTF-8, so I would use `str::from_utf8` here and avoid making a second copy of the string.

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 3]
```diff
@@ -1868,7 +1868,10 @@ extern "C" {
     pub fn LLVMRustGetLastError() -> *const c_char;
 
     /// Print the pass timings since static dtors aren't picking them up.
-    pub fn LLVMRustPrintPassTimings();
+    pub fn LLVMRustPrintPassTimings(size: *const usize) -> *const c_char;
```
Comment (oli-obk) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1871]: Nothing else in this file uses `usize`, please use `size_t` to make sure we don't rely on funny Rust<->C differences

File: compiler/rustc_codegen_llvm/src/lib.rs [hunk 4]
```diff
@@ -177,9 +177,34 @@ impl WriteBackendMethods for LlvmCodegenBackend {
...
-            llvm::LLVMRustPrintPassTimings();
-        }
+        let msg = unsafe {
+            let mut size = 0;
+            let cstr = llvm::LLVMRustPrintPassTimings(&mut size as *mut usize);
+            if cstr.is_null() {
+                "failed to get pass timings".into()
+            } else {
+                let timings = std::slice::from_raw_parts(cstr as *const u8, size);
+                let timings = std::str::from_utf8(timings).unwrap();
```
Comment (oli-obk) [inline compiler/rustc_codegen_llvm/src/lib.rs:187]: This creates a slice that refers to the data from `cstr`. The next line frees this data, making the Rust slice dangling.

Comment (oli-obk) [reply compiler/rustc_codegen_llvm/src/lib.rs:187]: Also don't even need utf8 validation, you could just write the bytes directly to stdout:

```rust
std::io::stdout().write_all(timings).unwrap();
```

this way you don't have the issue of preserving the data beyond the `free`

File: compiler/rustc_codegen_llvm/src/lib.rs [hunk 5]
```diff
@@ -177,9 +177,34 @@ impl WriteBackendMethods for LlvmCodegenBackend {
...
+    fn print_statistics(&self) {
+        let msg = unsafe {
+            let mut size = 0;
+            let cstr = llvm::LLVMRustPrintStatistics(&mut size as *mut usize);
+            if cstr.is_null() {
+                "failed to get stats".into()
+            } else {
+                let stats = std::slice::from_raw_parts(cstr as *const u8, size);
+                let stats = std::str::from_utf8(stats).unwrap();
+                libc::free(cstr as *mut _);
```
Comment (oli-obk) [inline compiler/rustc_codegen_llvm/src/lib.rs:203]: same UB here

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Review by khei4 — COMMENTED** [review]
Thank you for the review!

File: compiler/rustc_codegen_llvm/src/lib.rs [hunk 6]
```diff
@@ -178,7 +179,28 @@ impl WriteBackendMethods for LlvmCodegenBackend {
     type ThinBuffer = back::lto::ThinBuffer;
     fn print_pass_timings(&self) {
         unsafe {
-            llvm::LLVMRustPrintPassTimings();
+            let mut size = 0;
+            let cstr = llvm::LLVMRustPrintPassTimings(&mut size as *mut usize);
+            if cstr.is_null() {
+                std::io::stdout().write_all("failed to get pass timings".as_bytes()).unwrap();
```
Comment (oli-obk) [inline compiler/rustc_codegen_llvm/src/lib.rs:185]: write_all is only necessary to skip utf8 validation below, here you can just print

```suggestion
                println!("failed to get pass timings");
```

Comment (oli-obk) [reply compiler/rustc_codegen_llvm/src/lib.rs]: There's the same situation below which you did not address yet

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 7]
```diff
@@ -111,9 +112,26 @@ extern "C" void LLVMRustSetNormalizedTarget(LLVMModuleRef M,
...
-  raw_fd_ostream OS(2, false); // stderr.
-  TimerGroup::printAll(OS);
+extern "C" const char *LLVMRustPrintPassTimings(size_t *Len) {
+  std::string buf;
+  raw_string_ostream SS(buf);
+  TimerGroup::printAll(SS);
+  SS.flush();
+  *Len = buf.length();
+  char *CStr = (char *)malloc(*Len);
+  strcpy(CStr, buf.c_str());
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:122]: This should use memcpy, not strcpy.

**Review by nikic — APPROVED** [review]

**Pull request merged**